libkbfs/chat_local.go

```diff
@@ -6,74 +6,247 @@ package libkbfs
+			byID = make(convLocalByIDMap)
+			c.data.convs[tlfName][tlfType] = byID
+		}
+	}
+	for _, conv := range byID {
+		if conv.chanName == channelName {
+			return conv.convID, nil
+		}
+	}
+	// Make a new conversation.
+	var idBytes [8]byte
+	err := kbfscrypto.RandRead(idBytes[:])
+	if err != nil {
+		return nil, err
+	}
+	id := chat1.ConversationID(idBytes[:])
+	conv := &convLocal{
+		convID:   id,
+		chanName: channelName,
+	}
+	c.data.convs[tlfName][tlfType][id.String()] = conv
+	c.data.convsByID[id.String()] = conv
+	return id, nil
 }
 
 // SendTextMessage implements the Chat interface.
-func (c *ChatLocal) SendTextMessage(
+func (c *chatLocal) SendTextMessage(
 	ctx context.Context, tlfName tlf.CanonicalName, tlfType tlf.Type,
 	convID chat1.ConversationID, body string) error {
-	c.log.CDebugf(ctx, "Asked to send text message for %s, %s, "+
-		"but ignoring: %s", tlfName, tlfType, body)
+	c.data.lock.Lock()
+	defer c.data.lock.Unlock()
+	conv, ok := c.data.convs[tlfName][tlfType][convID.String()]
+	if !ok {
+		return errors.Errorf("Conversation %s doesn't exist", convID.String())
+	}
+	conv.messages = append(conv.messages, body)
+	conv.mtime = c.config.Clock().Now()
+	for _, cb := range conv.cbs {
+		cb(convID, body)
+	}
 	return nil
 }
 
+type chatHandleAndTime struct {
+	h     *TlfHandle
+	mtime time.Time
+}
+
 // GetGroupedInbox implements the Chat interface.
-func (c *ChatLocal) GetGroupedInbox(
+func (c *chatLocal) GetGroupedInbox(
 	ctx context.Context, chatType chat1.TopicType, maxChats int) (
-	results []tlf.CanonicalName, err error) {
-	return nil, nil
+	results []*TlfHandle, err error) {
+	if chatType != chat1.TopicType_KBFSFILEEDIT {
+		panic(fmt.Sprintf("Bad topic type: %d", chatType))
+	}
+
+	session, err := c.config.KBPKI().GetCurrentSession(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var handlesAndTimes []chatHandleAndTime
+
+	c.data.lock.Lock()
+	defer c.data.lock.Unlock()
+	for name, byType := range c.data.convs {
+		for t, byID := range byType {
+			if t == tlf.Public && string(name) != string(session.Name) {
+				// Skip public TLFs that aren't our own.
+				continue
+			}
+
+			h, err := GetHandleFromFolderNameAndType(
+				ctx, c.config.KBPKI(), c.config.MDOps(), string(name), t)
+			if err != nil {
+				return nil, err
+			}
+
+			// Only include if the current user can read the folder.
+			isReader, err := isReaderFromHandle(
+				ctx, h, c.config.KBPKI(), session.UID)
+			if err != nil {
+				return nil, err
+			}
+			if !isReader {
+				continue
+			}
+
+			hAndT := chatHandleAndTime{h: h}
+			for _, conv := range byID {
+				if conv.mtime.After(hAndT.mtime) {
+					hAndT.mtime = conv.mtime
+				}
+			}
+			handlesAndTimes = append(handlesAndTimes, hAndT)
+		}
+	}
+
+	sort.Slice(handlesAndTimes, func(i, j int) bool {
+		// Reverse sort so newest conversation is at index 0.
+		return handlesAndTimes[i].mtime.After(handlesAndTimes[j].mtime)
+	})
+	for i := 0; i < maxChats && i < len(handlesAndTimes); i++ {
```

Comment:
I'd prefer `i < len(handlesAndTimes)` to be first, since it's the primary range we care about. `maxChats` is just used to clip.